lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1609,6 +1609,135 @@ static void optimizeQuantizedMaxSplat(Function *F) {
   }
 }
 
+/// \returns A value representing the given \p constant converted
+/// to the destination type \p dstTy. If the conversion is not
+/// possible, this method returns NodeValue().
+static NodeValue convertConstant(Module &mod, Constant &constant,
+                                 TypeRef dstTy) {
+  // Sort out the easy case first.
+  if (constant.getType() == dstTy) {
+    return NodeValue(&constant, 0);
+  }
+  auto modifyConstantTyAndGet = [&]() -> Constant & {
+    Constant *oneUseCst = getUniquelyUsedConstant(&mod, constant);
+    assert(oneUseCst &&
+           "We should always be able to get a constant from a constant!");
+    // This type setting updates the type of the node.
+    // The underlying tensor still needs to be converted after this call.
+    oneUseCst->setType(0, dstTy);
+    return *oneUseCst;
+  };
+  const Tensor &tensor = constant.getPayload();
+  switch (tensor.getElementType()) {
+  case ElemKind::FloatTy:
+  case ElemKind::Float16Ty:
+    switch (dstTy->getElementType()) {
+    case ElemKind::FloatTy:
+    case ElemKind::Float16Ty: {
+      // Plain conversion: {FloatTy, Float16Ty} -> {FloatTy, Float16Ty}.
+      Constant &constantToBeModified = modifyConstantTyAndGet();
+      constantToBeModified.getPayload().convertToType(dstTy->getElementType());
+      return NodeValue(&constantToBeModified, 0);
+    }
+    case ElemKind::Int8QTy: {
+      // Quantization: {FloatTy, Float16Ty} -> Int8QTy.
+      Constant &constantToBeModified = modifyConstantTyAndGet();
+      TensorQuantizationParams params{dstTy->getScale(), dstTy->getOffset()};
+      Tensor &tensorToBeModified = constantToBeModified.getPayload();
+      // Right now we only quantize fp32 value.
+      // Add an assert on that, so that if it changes, we adapt the
+      // following code. Adapting the code would required to
+      // teach quantizeTensor how to deal with Float16Ty.
+      assert(tensor.getElementType() == ElemKind::FloatTy &&
+             "Type quantization not implemented");
+      tensorToBeModified =
+          quantization::quantizeTensor(tensorToBeModified, params);
+      return NodeValue(&constantToBeModified, 0);
+    }
+    default:
+      // Right now we only use int8 for quantization and fp32/fp16 conversions.
+      // Quantization: {FloatTy, Float16Ty} -> Int[16|32]QTy.
+      // Plain conversion: {FloatTy, Float16Ty} -> Int64ITy.
+      return NodeValue();
+    }
+  default:
+    // For now we don't see other quantize, dequantize, or rescale nodes
+    // directly attached to constants.
+    // Thus don't add code that will never be executed.
+    // Dequantization: Int[8|16|32]QTy -> {FloatTy, Float16Ty, Int64I}.
+    // Rescale: Int[8|16|32]QTy -> Int[8|16|32]QTy.
+    // Plain conversion: Int64ITy -> {FloatTy, Float16Ty}.
+    // Quantization: Int64ITy -> Int[8|16|32]QTy.
+    return NodeValue();
+  }
+}
+
+/// Optimize away ConvertToNode.
+/// This basically turns conversion(conversion A to B) to C"
```

Comment:
Small typo: missing opening `"` in `conversion(conversion A to B) to C"`

Reply:
Fixed the typo in PR #1869.